### PR TITLE
Unregister on event

### PR DESCRIPTION
**Pull request by tony-tran**

### CODE REVIEW

File: front/services/api/ReqwestApiClient.js [hunk 1]
```diff
@@ -158,6 +158,25 @@ export default class ReqwestApiClient {
         });
     }
 
+    unregisterEvent(eventSlug, token, callback) {
```
Comment (JeremieSamson) [inline front/services/api/ReqwestApiClient.js]: File is named "ReqwestApiClient", should be RequestApiClient no ?

File: src/Controller/EnMarche/EventController.php [hunk 2]
```diff
@@ -173,4 +174,28 @@ public function invitationSentAction(Request $request, Event $event): Response
...
     }
+
+    /**
+     * @Route("/desinscription", name="app_event_unregistration", condition="request.isXmlHttpRequest()")
+     * @Method("GET|POST")
+     */
+    public function unregistrationAction(Request $request, Event $event): JsonResponse
+    {
+        if (!$this->isCsrfTokenValid('citizen_action.unregistration', $token = $request->request->get('token'))) {
+            throw $this->createAccessDeniedException('Invalid CSRF protection token to unregister from the citizen action.');
```
Comment (ottaviano) [inline src/Controller/EnMarche/EventController.php]: I think BadRequestHttpException is more consistent in this case

File: src/Controller/EnMarche/EventController.php [hunk 3]
```diff
@@ -173,4 +174,28 @@ public function invitationSentAction(Request $request, Event $event): Response
...
+        }
+
+        $eventRegistrationManager = $this->get('app.event.registration_manager');
```
Comment (ottaviano) [inline src/Controller/EnMarche/EventController.php]: can you inject to the method this service?

File: src/Entity/BaseEvent.php [hunk 4]
```diff
@@ -238,7 +238,9 @@ public function incrementParticipantsCount(int $increment = 1): void
 
     public function decrementParticipantsCount(int $increment = 1): void
     {
-        $this->participantsCount -= $increment;
+        if (0 < $this->participantsCount) {
```
Comment (ottaviano) [inline src/Entity/BaseEvent.php]: good but not perfect, with `$this->participantsCount = 1` and `$increment = 2` it'll be negatif.

File: src/Twig/EventRuntime.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    private $eventRegistrationRepository;
+
+    public function __construct(EventRegistrationRepository $eventRegistrationRepository)
+    {
+        $this->eventRegistrationRepository = $eventRegistrationRepository;
+    }
+
+    public function isAlreadyParticipating(Event $event, ?UserInterface $user): bool
+    {
+        if (!$user instanceof Adherent || !$user) {
```
Comment (ottaviano) [inline src/Twig/EventRuntime.php]: the check `!$user instanceof Adherent` is enough here, no?

File: src/Twig/EventExtension.php [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+class EventExtension extends AbstractExtension
+{
+    public function getFunctions()
+    {
+        return [
+            new TwigFunction('is_already_participating', [EventRuntime::class, 'isAlreadyParticipating']),
```
Comment (ottaviano) [inline src/Twig/EventExtension.php]: what about `is_event_already_participating`

File: templates/events/_header.html.twig [hunk 7]
```diff
@@ -35,7 +35,11 @@
                 L'événement à été annulé
             </div>
         {% elseif not event.full%}
```
Comment (ottaviano) [inline templates/events/_header.html.twig]: can you add one space here pls?

File: src/Controller/EnMarche/EventController.php [hunk 8]
```diff
@@ -173,4 +174,28 @@ public function invitationSentAction(Request $request, Event $event): Response
...
         ]);
     }
+
+    /**
+     * @Route("/desinscription", name="app_event_unregistration", condition="request.isXmlHttpRequest()")
+     * @Method("GET|POST")
+     */
+    public function unregistrationAction(Request $request, Event $event): JsonResponse
+    {
+        if (!$this->isCsrfTokenValid('citizen_action.unregistration', $token = $request->request->get('token'))) {
```
Comment (ottaviano) [inline src/Controller/EnMarche/EventController.php]: what about renaming `citizen_action.unregistration` on `event.unregistration` for both CitizenAction and Event unsubscriptions? (as CitizenAction is an Event too)

File: templates/events/_header.html.twig [hunk 9]
```diff
@@ -35,7 +35,11 @@
...
-            <a class="btn btn--no-border {{ isFull?' btn--full':''}} register-event" href="{{ path("app_#{event.type}_attend", {slug: event.slug}) }}">Je veux participer</a>
+            {% if is_already_participating(event, app.user) %}
+                <button class="btn btn--no-border {{ isFull?' btn--full':''}} unregister-event" data-csrf-token="{{ csrf_token('citizen_action.unregistration') }}">Je ne veux plus participer</button>
```
Comment (ottaviano) [inline templates/events/_header.html.twig]: I think it's the wrong place for this button, because we are in `elseif not event.full`, when the event will be full I'll not see this button.

File: templates/search/_events.html.twig [hunk 10]
```diff
@@ -58,7 +58,7 @@
                 {% elseif not event.full %}
                     <a href="{{ path("app_#{event.type}_show", { slug: event.slug }) }}"
                        class="btn btn--blue" target="_blank" rel="noopener noreferrer">
-                        Participer
+                        Voir plus
```
Comment (ottaviano) [inline templates/search/_events.html.twig]: why this change?

File: front/services/api/ReqwestApiClient.js (hunk 1 above)
Comment (tony-tran) [reply front/services/api/ReqwestApiClient.js]: Yes, I catch that, I propose to change this in another PR WDYT ?

File: src/Twig/EventRuntime.php (hunk 5 above)
Comment (tony-tran) [reply src/Twig/EventRuntime.php]: Yes because $user can be null (unauthenticated user)

File: templates/search/_events.html.twig (hunk 10 above)
Comment (tony-tran) [reply templates/search/_events.html.twig]: It's a P.O demand

File: src/Twig/EventRuntime.php (hunk 5 above)
Comment (ottaviano) [reply src/Twig/EventRuntime.php]: @tony-tran so if yes, and the first check is enough you will remove the second one?

File: src/Controller/EnMarche/EventController.php [hunk 11]
```diff
@@ -173,4 +176,26 @@ public function invitationSentAction(Request $request, Event $event): Response
...
+     * @Method("GET|POST")
+     */
+    public function unregistrationAction(Request $request, Event $event, EventRegistrationManager $eventRegistrationManager): JsonResponse
+    {
+        if (!$this->isCsrfTokenValid('event.unregistration', $token = $request->request->get('token'))) {
+            throw new BadRequestHttpException('Invalid CSRF protection token to unregister from the citizen action.');
+        }
+
+        if (!($adherentEventRegistration = $eventRegistrationManager->searchRegistration($event, $this->getUser()->getEmailAddress(), null))) {
+            return new JsonResponse(
```
Comment (ottaviano) [inline src/Controller/EnMarche/EventController.php]: what about:
```php
throw $this->createNotFoundException(
    'Impossible d\'exécuter la désinscription de l\'évènement, votre inscription n\'est pas trouvée.'
)
```

File: src/Twig/EventRuntime.php [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        $this->eventRegistrationRepository = $eventRegistrationRepository;
+    }
+
+    public function isEventAlreadyParticipating(Event $event, ?UserInterface $user): bool
+    {
+        if (!$user instanceof Adherent) {
+            return false;
+        }
+
+        return $this->eventRegistrationRepository->findGuestRegistration($event->getUuid(), $user->getEmailAddress()) ? true : false;
```
Comment (ottaviano) [inline src/Twig/EventRuntime.php]: as an alternative:
```
return (bool) $this->eventRegistrationRepository->findGuestRegistration(
    $event->getUuid(), 
    $user->getEmailAddress()
);
```

File: templates/events/_header.html.twig [hunk 13]
```diff
@@ -34,12 +34,20 @@
             <div class="text--white text--center">
                 L'événement à été annulé
             </div>
-        {% elseif not event.full%}
-            <a class="btn btn--no-border {{ isFull?' btn--full':''}} register-event" href="{{ path("app_#{event.type}_attend", {slug: event.slug}) }}">Je veux participer</a>
+        {% elseif not event.full %}
```
Comment (ottaviano) [inline templates/events/_header.html.twig]: can you refactor this part for writing once the button block, something like that: 
```
{% else %}
     {% if is_event_already_participating(event, app.user) %}
          <button .../>
     {% endif %}

     {% if not event.full %}
           // ...
     {% else %}
           // ...
     {% endif %}
{% endif %}
```

File: src/Twig/EventRuntime.php (hunk 12 above)
Comment (tony-tran) [reply src/Twig/EventRuntime.php]: I prefer don't cast the result to be sure to control the return.

Comment (ottaviano) [reply src/Twig/EventRuntime.php]: it is the exactly what you are doing with 
```
return $object ? true : false;
=== 
return (bool) object;
```
to avaluate the result of `if ($object)` PHP cast the $object in boolean value.
